src/libopensc/pkcs15-sec.c

```diff
@@ -287,7 +287,15 @@ int sc_pkcs15_derive(struct sc_pkcs15_card *p15card,
 
 	/* If card stores derived key on card, then no data is returned
 	 * and the key must be used on the card. */
-	*poutlen = r;
+	
+	switch (obj->type) {
+		case SC_PKCS15_TYPE_PRKEY_EC:
+	/* But in case the card returned the shared secret, don't overwrite *poutlen */
+			break;
+		default:
+			*poutlen = r;
+	}
+	
```

Comment:
This sections of the code is not needed. 
The use_key in line 277 should be returning the length of the derived data the card returned, 
or 0 if the card did not return any data. if r<0 indicates an error and returned at 278.
If card returned rsa padded data  the padding is removed and r would be the new length. 

So in the original code:
*poutlen = r;

This will be setting *poutlen to the size of the data returned - any padding removed. 
if the the card as in the EC case, there is no padding , and if the card returned data, *poutlen would already be the same as r. 

Please remove this section as it makes it more confusing.